File: ios/PaddleupShared/Sources/PaddleupShared/Geo.swift
import Foundation

public enum Geo {
    /// Haversine distance in meters between two WGS-84 coordinates.
    public static func haversineMeters(_ lat1: Double, _ lon1: Double, _ lat2: Double, _ lon2: Double) -> Double {
        let r = 6_371_000.0
        let φ1 = lat1 * .pi / 180
        let φ2 = lat2 * .pi / 180
        let dφ = (lat2 - lat1) * .pi / 180
        let dλ = (lon2 - lon1) * .pi / 180
        let a = sin(dφ / 2) * sin(dφ / 2) + cos(φ1) * cos(φ2) * sin(dλ / 2) * sin(dλ / 2)
        return 2 * r * asin(min(1.0, sqrt(a)))
    }

    /// Stride-based downsample to at most `target` items.
    public static func downsample<T>(_ points: [T], target: Int) -> [T] {
        guard points.count > target, target > 0 else { return points }
        let step = Double(points.count) / Double(target)
        var out = (0..<target).map { points[Int(Double($0) * step)] }
        if let last = points.last { out.append(last) }
        return out
    }
}

File: src/services/geo.ts
/** Haversine distance in meters between two WGS-84 coordinates. */
export function haversineMeters(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const r = 6_371_000;
  const phi1 = (lat1 * Math.PI) / 180;
  const phi2 = (lat2 * Math.PI) / 180;
  const dPhi = ((lat2 - lat1) * Math.PI) / 180;
  const dLam = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dPhi / 2) ** 2 + Math.cos(phi1) * Math.cos(phi2) * Math.sin(dLam / 2) ** 2;
  return 2 * r * Math.asin(Math.min(1, Math.sqrt(a)));
}

/** Stride-based downsample to at most `target` items. */
export function downsample<T>(points: T[], target: number): T[] {
  if (points.length <= target || target <= 0) return points;
  const step = points.length / target;
  const out: T[] = [];
  for (let i = 0; i < target; i++) out.push(points[Math.floor(i * step)]!);
  if (points.length > 0) out.push(points[points.length - 1]!);
  return out;
}
